apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/StorageClientImplRetryTest.cs

```diff
@@ -0,0 +1,459 @@
+using Object = Google.Apis.Storage.v1.Data.Object;
+
+namespace Google.Cloud.Storage.V1.Tests
+{
+    public class StorageClientImplRetryTest
+    {
+        #region Bucket test cases
+
+        [Fact]
+        public void GetBucket_NoRetry() => NoRetryHelper(service => service.Buckets.Get("bucket"), client => client.GetBucket("bucket"));
+
+        [Fact]
+        public void GetBucket_RetryOnce() => RetryOnceHelper(service => service.Buckets.Get("bucket"), client => client.GetBucket("bucket"));
+
+        [Fact]
+        public void GetBucket_RetryThenFail() => RetryThenFailHelper(service => service.Buckets.Get("bucket"), client => client.GetBucket("bucket"));
+
+        [Fact]
+        public void DeleteBucket_NoRetry() => NoRetryHelper(service => service.Buckets.Delete("bucket"), client => client.DeleteBucket("bucket"));
+
+        [Fact]
+        public void DeleteBucket_RetryOnce() => RetryOnce(service => service.Buckets.Delete("bucket"), client => client.DeleteBucket("bucket"), "check");
+
+        [Fact]
+        public void DeleteBucket_RetryThenFail() => RetryThenFail(service => service.Buckets.Delete("bucket"), client => client.DeleteBucket("bucket"), "check");
+
+        [Fact]
+        public void ListBucket_NoRetry() => NoRetryHelper(service => new BucketsResource.ListRequest(service, "project"), client => client.ListBuckets("project").Count());
+
+        [Fact]
+        public void ListBucket_RetryOnce() => RetryOnceHelper(service => service.Buckets.List("project"), client => client.ListBuckets("project").Count());
+
+        [Fact]
+        public void ListBucket_RetryThenFail() => RetryThenFailHelper(service => service.Buckets.List("project"), client => client.ListBuckets("project").Count());
+
+        [Fact]
+        public void InsertBucket_RetryOnce() => RetryOnceHelper(
+            service => service.Buckets.Insert(new Bucket { Name = "bucket" }, "project"), 
+            client => client.CreateBucket("project", "bucket"));
+
+        [Fact]
+        public void InsertBucket_RetryThenFail() => RetryThenFailHelper(
+            service => service.Buckets.Insert(new Bucket { Name = "bucket" }, "project"), 
+            client => client.CreateBucket("project", "bucket"));
+
+        [Fact]
+        public void InsertBucket_NoRetry() => NoRetryHelper(
+            service => service.Buckets.Insert(new Bucket { Name = "bucket" }, "project"), 
+            client => client.CreateBucket("project", "bucket"));
+
+        [Fact]
+        public void UpdateBucket_WithMetageneration_RetryOnce() => RetryOnceHelper(
+            service => new BucketsResource.UpdateRequest(service, new Bucket { Name = "bucket" }, "bucket") { IfMetagenerationMatch = 70 }, 
+            client => client.UpdateBucket(new Bucket { Name = "bucket" }, new UpdateBucketOptions() { IfMetagenerationMatch = 70 }));
+
+        [Fact]
+        public void UpdateBucket_WithoutMetgeneration_NoRetry() => NoRetryHelper(
```

Comment:
This is another "Metgeneration" - it looks like there are still three of them...